### PR TITLE
Sync wasi-filesystem with wit definitions in standards repo

### PR DESCRIPTION
**Pull request by pchickey**
This PR pulls in the latest wit files from the wasi-filesystem repo, with the addition of just one pending upstream change to remove the nonblocking descriptor-flag.

The changes look like:
* the `wasi:filesystem/filesystem` interface has been renamed `wasi:filesystem/types`. This resulted in lots of renaming across the repo. For the main filesystem impl I use the new `types::` module prefix but in most other places I followed `use ...::filesystem::types as filesystem;` to make it clearer to read.
* the interface removes `device` and `inode` from the descriptor-stat struct, and provides new `metadata-hash` family functions to replace this concept.
    * these new methods are implemented in preview2 host, using `std::hash::Hasher` to hash the device and inode of the file.
    * the preview 1 adapters still need to fill in a `device` and `inode` in some places (`fd_filestat_get` and `fd_readdir`). I elected to fill in `1` for the `device` (because the stdio stream "files" use `0`) and use the lower 64 bits of the metadata hash to fill in the `inode`.
    * the preview 1 host adapter needed to use an `async` method to calculate each `inode` in `fd_readdir`, so I had to move some calculation out of the iterator map and into a for loop, done eagerly for each entry in the directory. Ideally this would be lazy for only entries in the directory that are being serialized as part of the adapter, or cache it between entries, but I didn't consider this worthwhile to invest in right now.
* the new wasi:filesystem/preopens interface replaces the local wasi:cli-base/preopens. (There is no such thing as an upstream cli-base package, but morally it is in wasi-cli)

See also:
https://github.com/WebAssembly/wasi-filesystem/pull/125: remove nonblocking descriptor flag. still open.
https://github.com/WebAssembly/wasi-filesystem/pull/126: fix parameters to metadata functions. merged.

### CODE REVIEW

File: crates/wasi/src/preview2/mod.rs [hunk 1]
```diff
@@ -49,12 +49,12 @@ pub mod bindings {
...
               import wasi:poll/poll
               import wasi:io/streams
-              import wasi:filesystem/filesystem
+              import wasi:filesystem/types
             ",
                 tracing: true,
                 trappable_error_type: {
                     "streams"::"stream-error": Error,
-                    "filesystem"::"error-code": Error,
+                    "types"::"error-code": Error,
```
Comment (elliottt) [inline crates/wasi/src/preview2/mod.rs:57]: These will need to be updated after #6795 merges.

File: crates/wasi/src/preview2/preview2/filesystem.rs [hunk 2]
```diff
@@ -837,10 +828,112 @@ impl<T: WasiView> filesystem::Host for T {
...
+    use cap_fs_ext::MetadataExt;
+    // Without incurring any deps, std provides us with a 64 bit hash
+    // function:
+    use std::hash::{BuildHasher, Hasher};
+    // We only want to initialize the BuildHasher impl once so that each Hasher we create performs
+    // the same calculation.
+    // Note that this means that the metadata hash (which becomes a preview1 ino) will be different
+    // in each wasmtime process!
+    static BUILD_HASHER: std::sync::OnceLock<std::collections::hash_map::RandomState> =
+        std::sync::OnceLock::new();
```
Comment (elliottt) [inline crates/wasi/src/preview2/preview2/filesystem.rs:915]: Why not use `std::collections::hash_map::DefaultHasher` here instead?

**Review by elliottt — COMMENTED** [review]
I think this looks great! Just had a question about the hasher choice before merging.

File: crates/wasi/src/preview2/preview2/filesystem.rs (hunk 2 above)
Comment (pchickey) [reply crates/wasi/src/preview2/preview2/filesystem.rs:915]: I missed it, but that is a better option.

**Review by elliottt — APPROVED** [review]
:rocket:

**Pull request merged**